### PR TITLE
Move TIL image decoding to a common place

### PR DESCRIPTION
**Pull request by ihhub**
as discussed in #6601

### CODE REVIEW

File: src/engine/image_tool.cpp [hunk 1]
```diff
@@ -325,6 +325,24 @@ namespace fheroes2
...
     }
 
+    void decodeTILImages( const uint8_t * data, const size_t imageCount, const int32_t width, const int32_t height, std::vector<Image> & output )
+    {
+        assert( data != nullptr && imageCount > 0 && width > 0 && height > 0 );
+
+        output.clear();
+        output.resize( imageCount );
+
+        const size_t imageSize = static_cast<size_t>( width * height );
```
Comment (oleg-derevenetz) [inline src/engine/image_tool.cpp:335]: We can avoid this warning this way:

```suggestion
        const size_t imageSize = static_cast<size_t>( width ) * height;
```

File: src/fheroes2/agg/agg_image.cpp [hunk 2]
```diff
@@ -3421,24 +3421,15 @@ namespace fheroes2
...
 
-                const uint32_t count = buffer.getLE16();
-                const uint32_t width = buffer.getLE16();
-                const uint32_t height = buffer.getLE16();
-                const uint32_t size = width * height;
-                if ( headerSize + count * size != data.size() ) {
+                const int32_t count = buffer.getLE16();
+                const int32_t width = buffer.getLE16();
+                const int32_t height = buffer.getLE16();
+                if ( count < 1 || width < 1 || height < 0 || ( headerSize + static_cast<size_t>( count * width * height ) ) != data.size() ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/agg/agg_image.cpp:3427]: Why `width < 1` but `height < 0`? Are there valid images with `width == 1` and `height == 0`? Also `getLE16()` returns `uint16_t`, so it is impossible for `uint16_t` to become negative on widening conversion to `int32_t`.

Also I propose to replace `static_cast<size_t>( count * width * height )` by `static_cast<size_t>( count ) * width * height` here as well.

File: src/fheroes2/agg/agg_image.cpp [hunk 3]
```diff
@@ -3447,7 +3438,7 @@ namespace fheroes2
                     const bool horizontalFlip = ( shapeId & 2 ) != 0;
                     const bool verticalFlip = ( shapeId & 1 ) != 0;
 
-                    for ( uint32_t i = 0; i < count; ++i ) {
+                    for ( int32_t i = 0; i < count; ++i ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/agg/agg_image.cpp:3441]: Since `i` is used just as an index of `vector` here, I propose to make it `size_t`.

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]
Hi @ihhub I just left a few proposals and questions, could you please take a look when you have time?

File: src/fheroes2/agg/agg_image.cpp (hunk 3 above)
Comment (ihhub) [reply src/fheroes2/agg/agg_image.cpp:3441]: But count is **int32_t**.

Comment (ihhub) [reply src/fheroes2/agg/agg_image.cpp:3441]: Actually, I will reduce it to normal uint16_t as we read it from the stream.

Comment (oleg-derevenetz) [reply src/fheroes2/agg/agg_image.cpp:3441]: > But count is int32_t.

Ah yes, it is signed now.

**Review by oleg-derevenetz — APPROVED** [review]
LGTM. As soon as IWYU become happy, let's try to use it in the new `til2img` tool.

**Pull request merged**